### PR TITLE
API updates for v2 `meta` arguments and types

### PR DESCRIPTION
**Pull request by chaance**
This PR does a few things we'll want to include behind the `v2_meta` flag before the major release.

- `V2_HtmlMetaDescriptor` has been renamed to `V2_MetaDescriptor`. This is shorter but also more aligned with the `links` function return value type, which is `LinkDescriptor`. It also makes a bit more sense as we support values for other HTML tags than `<meta>`.
- `V2_HtmlMetaDescriptor` and `V2_HtmlMetaFunction` types should be imported from `@remix-run/react` instead of `@remix-run/<runtime>` since A) the primary responsibility of `meta` is rendering, and B) `matches` data only makes sense in the context of an app calling `useMatches`. I think we should follow suit here with `LinksFunction` in v2, but that's for a separate PR.
- The `meta` function's arguments have been simplified
  - `parentsData` has been removed, as each route's loader data is available on the `data` property of its respective `match` object
    ```tsx
    // before
    export function meta({ parentsData }) {
      return {
        title: parentsData['routes/some-route'].title,
      };
    };
    // after
    export function meta({ matches }) {
      return {
        title: matches.find((match) => match.id === 'routes/some-route').data.title,
      };
    };
    ```
  - The `route` property on route matches has been removed, as relevant match data is attached directly to the match object
    ```tsx
    // before
    export function meta({ parentsData }) {
      let rootModule = matches.find(match => match.route.id === "root");
    };
    // after
    export function meta({ matches }) {
      let rootModule = matches.find(match => match.id === "root");
    };
    ```

### CODE REVIEW

File: docs/route/meta.md [hunk 1]
```diff
@@ -150,7 +150,7 @@ export const meta: V2_MetaFunction = () => {
 };
 ```
 
-Meta functions return an array of `V2_HtmlMetaDescriptor` objects. These objects map one-to-one with normal HTML meta tags:
+Meta functions return an array of `V2_MetaDescriptor` objects. These objects map one-to-one with normal HTML meta tags:
```
Comment (ryanflorence) [inline docs/route/meta.md:153]: > normal HTML tags

since we have links/script too

File: packages/remix-react/components.tsx [hunk 2]
```diff
@@ -568,12 +571,11 @@ function V1Meta() {
     if (routeModule.meta) {
       let routeMeta =
         typeof routeModule.meta === "function"
-          ? routeModule.meta({
+          ? (routeModule.meta as V1_MetaFunction)({
```
Comment (ryanflorence) [inline packages/remix-react/components.tsx:574]: shouldn't this type already come from the routeModule type? we don't have to cast the other types of this object, why is this property special?

File: packages/remix-react/components.tsx [hunk 3]
```diff
@@ -651,39 +653,28 @@ function V1Meta() {
...
-
-    let routeMeta: V2_HtmlMetaDescriptor[] | V1_HtmlMetaDescriptor | undefined =
-      [];
+    let routeMeta: V2_MetaDescriptor[] | V1_HtmlMetaDescriptor | undefined = [];
 
     if (routeModule?.meta) {
       routeMeta =
         typeof routeModule.meta === "function"
-          ? routeModule.meta({
+          ? (routeModule.meta as V2_MetaFunction)({
```
Comment (ryanflorence) [inline packages/remix-react/components.tsx:673]: same question here as above

**Review by ryanflorence — APPROVED** [review]
just some questions, nothing blocking though.

File: packages/remix-react/components.tsx (hunk 2 above)
Comment (chaance) [reply packages/remix-react/components.tsx:574]: Because it's a union type on the route module, TS doesn't know how to disambiguate at the call-site.

**Pull request merged**